bookdown/02-basics-resampling.Rmd

```diff
@@ -180,3 +180,169 @@ autoplot(rr, type = "roc")
+
+
+- **Train Set**: Used to train ML algorithms and preprocessing methods.
+
+- **Validation Set**: Used to compare and choose hyperparameters, early stopping etc.
+
+- **Test Set**: Used to obtain an un-biased estimate of the desired metric.
+
+The attentive reader has surely observed, that those concepts do not directly match to the way **mlr3** handles such data splitting procedures.
+As **mlr3** tries to stay general in order to offer a maximum of flexibility, simple splitting procedures are abstrated into **Resampling Methods** such as the ones obtained via `r ref("rsmp()")`.
+In order to tune and select hyperparameters, instead of using a **validation split**, we choose a **inner** resampling method that is used for this selection.
+This e.g. happens inside an `r ref("AutoTuner")` from `r mlr_pkg("mlr3tuning")`, where we provide an additional inner resampling method for tuning. If we now `r ref("resample()")` this `r ref("AutoTuner")`, we get the exact desired effect:
+The `r ref("AutoTuner")` internally tunes the  hyperparameters of our algorithm and `r ref("resample()")` provides us with an uniased estimate of this tuned algorithms performance.
+
+There are two approaches towards obtaining unbiased performance estimates from simple train-validation-test splits in **mlr3**.
+Either using a `r ref("AutoTuner")` together with `r ref("resample()")`, or using by keeping test data out using mlr3's `row_roles` mechanism.
+While the first approach is less flexible it is easy to understand and simple to apply. The second approach is slightly more complex, but also allows for a lot more flexibility and automation.
+
+
+#### How can I keep out test data for prediction?
+
+
+Let us assume we have the following pre-defined **train** and **test** `row_ids`.
+
+```{r}
+t = tsk("iris")
+train = c(1:30, 51:80, 101:130)
+test = c(31:50, 81:100, 131:150)
+```
+
+We can now look at the `task`'s `row_roles`
+
+```{r}
+t$row_roles
+```
+
+We can see, that there is one set for `r names(t$row_roles)[1]`  and one for `r names(t$row_roles)[2]`. We can now set our `test` set to `validation` in order to keep it out from future computations.
+
+```{r}
+t$set_row_role(test, "validation")
+```
+
+If we now create and instantiate a `holdout` resampling instance, we can see that
+those rows will not be used:
+
+```{r}
+rdesc = rsmp("holdout")$instantiate(t)
+sort(c(rdesc$train_set(1), rdesc$test_set(1)))
+```
+
+We are now going to use this instance to compare two versions of the `r ref("rpart")` learner with different hyperparameter settings, choose the better one and train this on the full data.
+
+```{r}
+resample(t, lrn("classif.rpart", cp = .1), rdesc)$score()
+resample(t, lrn("classif.rpart", cp = .5), rdesc)$score()
+
+# cp = 0.1 works better, train it and predict on validation rows.
+l = lrn("classif.rpart", cp = .1)$train(t)
+l$predict(t, row_ids = t$row_roles$validation)$score()
```

Comment:
use a measure that validates on validation data.

Reply:
You might need branch https://github.com/mlr-org/mlr3/tree/holdout :D

Reply:
Names are up for discussion